python/dgl/backend/pytorch/sparse_emb.py

```diff
@@ -0,0 +1,160 @@
+    optimizers in the future.
+
+    Parameters
+    ----------
+    num_embeddings : int
+        The number of embeddings. Currently, the number of embeddings has to be the same as
+        the number of nodes or the number of edges.
+    embedding_dim : int
+        The dimension size of embeddings.
+    name : str
+        The name of the embeddings. The name should uniquely identify embeddings in a system.
+    init_func : callable, optional
+        The function to create the initial data. If the init function is not provided,
+        the values of the embeddings are initialized to zero.
+    group : str, optional
+        Group name.
+    host_name : str, optional
+        The hostname or IP Address the server store should run on.
+    port : int, optional
+        The port on which the server store should listen for incoming requests.
+
+    Examples
+    --------
+    Before launching multiple gpu processes
+
+    >>> def initializer(emb):
+            th.nn.init.xavier_uniform_(emb)
+            return emb
+
+    In each multiple gpu process
+
+    >>> emb = dgl.GraphSparseEmbedding(g.number_of_nodes(), 10, 'emb', init_func=initializer)
+    >>> optimizer = dgl.SparseAdagrad([emb], lr=0.001)
+    >>> for blocks in dataloader:
+    ...     feats = emb(nids, gpu_0)
+    ...     loss = F.sum(feats + 1, 0)
+    ...     loss.backward()
+    ...     optimizer.step()
+    '''
+
+    def __init__(self, num_embeddings, embedding_dim, name,
+                 init_func=None, group=None, host_name='127.0.0.1', port=12346):
+        global _store
+
+        # Check whether it is multi-gpu training or not.
+        if th.distributed.is_initialized():
+            rank = th.distributed.get_rank() if group is None \
+                else th.distributed.get_rank(group)
+            world_size = th.distributed.get_world_size() if group is None \
+                else th.distributed.get_world_size(group)
+        else:
+            rank = -1
+            world_size = 0
+        self._rank = rank
+        self._world_size = world_size
+
+        if rank <= 0:
+            emb = create_shared_mem_array(name, (num_embeddings, embedding_dim), th.float32)
+            if init_func is not None:
+                emb = init_func(emb)
+        if rank == 0:
+            if world_size > 1:
+                # for multi-gpu training, setup a TCPStore for
+                # embeding status synchronization across GPU processes
+                if _store is None:
+                    _store = th.distributed.TCPStore(
+                        host_name, port, world_size, True, timedelta(seconds=30))
+                for i in range(1, world_size):
+                    # send embs
+                    _store.set(name, name)
+        elif rank > 0:
+            # receive
+            if _store is None:
+                _store = th.distributed.TCPStore(
+                    host_name, port, world_size, False, timedelta(seconds=30))
+            _store.wait([name])
```

Comment:
what is communicated by `TCPStore`? name?

Reply:
host name and port